File: 11/day11.pl
:- ['intcode.pl'].

parse(Input) :-
    csv_read_file("day11.input", [Row], []),
    Row =.. [_|Input].

turn_left(DX-DY, LX-LY) :-
    LX #= -DY,
    LY #= DX.

turn_right(DX-DY, LX-LY) :-
    LX #= DY,
    LY #= -DX.

left_or_right(0, DX-DY, NX-NY) :-
    turn_left(DX-DY, NX-NY).
left_or_right(1, DX-DY, NX-NY) :-
    turn_right(DX-DY, NX-NY).

camera(X-Y, Panels, Color) :-
    ( memberchk(X-Y-C, Panels)
    ->
        Color = C
    ;
        Color = 0
    ).

paint(X-Y, Color, PanelsIn, PanelsOut) :-
    ( selectchk(X-Y-_, PanelsIn, Rem)
    ->
        PanelsOut = [X-Y-Color | Rem]
    ;
        PanelsOut = [X-Y-Color | PanelsIn]
    ).

paint_loop(X-Y, DX-DY, ProgramConfig, PanelsIn, PanelsOut) :-
    ProgramConfig = [Input, Index, RelBase, Output],
    camera(X-Y, PanelsIn, Color),
    % TODO: intcode.give_input
    append(Input, [Color], NewInput),
    run(NewInput, RemInput, Index, NewIndex, RelBase, NewRelBase, Output, NewOutput, Halted),
    (
        Halted=true,
        PanelsIn = PanelsOut
    ;
        NewOutput = [SetColor, Direction | RestOutput],
        paint(X-Y, SetColor, PanelsIn, NewPanels),
        left_or_right(Direction, DX-DY, NDX-NDY),
        NX #= X + NDX,
        NY #= Y + NDY,
        NewProgramConfig = [RemInput, NewIndex, NewRelBase, RestOutput],
        paint_loop(NX-NY, NDX-NDY, NewProgramConfig, NewPanels, PanelsOut)
    ).

run_robot(Program, PanelsIn, PanelsOut) :-
    list_to_mem(Program),
    % TODO: intcode.new_program
    ProgramConfig = [ [], 0, 0, [] ],
    paint_loop(0-0, 0-1, ProgramConfig, PanelsIn, PanelsOut),
    clean_mem.

bounds(Panels, MinX-MinY, MaxX-MaxY) :-
    maplist([X-_-_, Z]>>(Z=X), Panels, Xs),
    min_list(Xs, MinX),
    max_list(Xs, MaxX),
    maplist([_-Y-_, Z]>>(Z=Y), Panels, Ys),
    min_list(Ys, MinY),
    max_list(Ys, MaxY).

print(_, _-A, _-Y) :- A #= Y+1.
print(Panels, MinX-MinY, MaxX-MaxY) :-
    Len #= MaxX - MinX + 1,
    length(Xs, Len),
    Xs ins MinX..MaxX,
    chain(Xs, #<),
    all_distinct(Xs),
    maplist({Panels, MaxY}/[In,Out]>>camera(In-MaxY, Panels, Out), Xs, Line),
    maplist([In, Out]>>(In=1->Out='\u2588';Out=' '), Line, Ascii),
    string_chars(Str, Ascii),
    writeln(Str),
    NMaxY #= MaxY - 1,
    print(Panels, MinX-MinY, MaxX-NMaxY).

run :-
    parse(Program),
    run_robot(Program, [], Panels),
    length(Panels, N),
    format("Part 1: ~w~n", [N]),
    run_robot(Program, [0-0-1], PanelsSingleWhite),
    bounds(PanelsSingleWhite, MinX-MinY, MaxX-MaxY),
    print(PanelsSingleWhite, MinX-MinY, MaxX-MaxY).

File: 11/day11.rs
use std::fs::File;
use std::io;
use std::io::Read;
use std::str::FromStr;
use std::collections::HashMap;

mod intcode;

fn parse(filename: &str) -> io::Result<HashMap<i64,i64>> {
    let mut file = File::open(filename)?;
    let mut input = String::new();
    file.read_to_string(&mut input)?;
    Ok(str_to_map(&input))
}

fn str_to_map(input: &str) -> HashMap<i64,i64> {
    let mut m = HashMap::new();
    for (i, x) in input.trim_end().split(',').map(|x| i64::from_str(&x).unwrap()).enumerate() {
        m.insert(i as i64, x);
    }
    m
}

#[derive(PartialEq,Eq,Hash,Debug,Clone,Copy)]
struct Coord {
    x: i32,
    y: i32
}

impl Coord {
    fn new(x: i32, y: i32) -> Coord {
        Coord{x:x, y:y}
    }

    fn turn_left(&mut self) {
        let (x,y) = (self.x, self.y);
        self.x = -y;
        self.y = x;
    }

    fn turn_right(&mut self) {
        let (x,y) = (self.x, self.y);
        self.x = y;
        self.y = -x;
    }
}

fn run_robot(brain: &HashMap<i64, i64>, panels: &mut HashMap<Coord, i64>) {
    let mut program = intcode::Program::new(&brain);
    let mut pos = Coord::new(0, 0);
    let mut direction = Coord::new(0, 1);
    while !program.halted() {
        let camera = panels.get(&pos).or(Some(&0)).unwrap();
        program.give_input(*camera);
        program.run();
        let color = program.get_output().unwrap();
        panels.insert(pos, color);
        match program.get_output().unwrap() {
            0 => direction.turn_left(),
            1 => direction.turn_right(),
            _ => panic!()
        };
        pos.x += direction.x;
        pos.y += direction.y;
    }
}

fn main() {
    let input = parse("day11.input").unwrap();
    let mut panels = HashMap::new();
    run_robot(&input, &mut panels);
    println!("Part 1: {}", panels.len());
    let mut panels_single_white = HashMap::new();
    panels_single_white.insert(Coord::new(0,0), 1);
    run_robot(&input, &mut panels_single_white);
    let mut mincoords = Coord::new(std::i32::MAX, std::i32::MAX); 
    let mut maxcoords = Coord::new(std::i32::MIN, std::i32::MIN); 
    for k in panels_single_white.keys() {
        if k.x < mincoords.x {
            mincoords.x = k.x;
        }
        if k.y < mincoords.y {
            mincoords.y = k.y;
        }
        if k.x > maxcoords.x {
            maxcoords.x = k.x;
        }
        if k.y > maxcoords.y {
            maxcoords.y = k.y;
        }
    }
    println!("Part 2: ");
    for y in (mincoords.y..maxcoords.y+1).rev() {
        for x in mincoords.x..maxcoords.x+1 {
            let camera = panels_single_white.get(&Coord::new(x,y)).or(Some(&0)).unwrap();
            match camera {
                0 => print!(" "),
                1 => print!("\u{2588}"),
                _ => panic!()
            }
        }
        println!("");
    }
}
